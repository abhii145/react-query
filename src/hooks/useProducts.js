import { keepPreviousData, useQuery } from "@tanstack/react-query"

const fetchCategories = async () => {
  const response = await fetch(`https://dummyjson.com/products/categories`)
  const data = await response.json()
  return data
}

const fetchProducts = async ({
  limit,
  skip,
  searchQuery,
  category,
  sortBy,
  order,
}) => {
  let url = `https://dummyjson.com/products/search?limit=${limit}&skip=${skip}&q=${searchQuery}&sortBy=${sortBy}&order=${order}`
  if (category && category !== "All") {
    url = `https://dummyjson.com/products/category/${category}?limit=${limit}&skip=${skip}&sortBy=${sortBy}&order=${order}`
  }

  const response = await fetch(url)
  const data = await response.json()
  return {
    products: data.products,
    total: data.total,
  }
}

export const useCategories = () => {
  return useQuery({
    queryKey: ["categories"],
    queryFn: fetchCategories,
    placeholderData: keepPreviousData,
    staleTime: 1000 * 60 * 5,
  })
}

export const useProducts = ({
  limit,
  skip,
  searchQuery,
  category,
  sortBy,
  order,
}) => {
  return useQuery({
    queryKey: ["products", limit, skip, searchQuery, category, sortBy, order],
    queryFn: () =>
      fetchProducts({ limit, skip, searchQuery, category, sortBy, order }),
    placeholderData: keepPreviousData,
    staleTime: 1000 * 60 * 5,
  })
}
