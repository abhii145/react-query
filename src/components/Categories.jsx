import { Link } from "react-router"
import { useCategories } from "../hooks/useProducts"
import { categoryIcons } from "../constants"
import Loading from "./Loading"

const Categories = () => {
  const { data: categories, isLoading } = useCategories()

  if (isLoading) {
    return <Loading />
  }

  return (
    <section>
      <div className="flex justify-between items-center mb-4">
        <h2 className="text-2xl font-bold">Categories</h2>
        <Link to="/products" className="text-blue-700 hover:underline text-sm">
          View All Products
        </Link>
      </div>
      <div className="grid grid-cols-2 sm:grid-cols-3 md:grid-cols-4 gap-4">
        {categories?.map((category) => (
          <Link
            to={`/products?category=${category.slug}`}
            key={category.slug}
            className="p-4 bg-white rounded-lg shadow hover:bg-gray-200 flex items-center gap-3"
          >
            <div className="text-blue-700 text-2xl">
              {categoryIcons[category.slug]}
            </div>
            <span className="font-medium">{category.name}</span>
          </Link>
        ))}
      </div>
    </section>
  )
}

export default Categories
