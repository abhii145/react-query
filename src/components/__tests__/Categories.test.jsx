// src/components/__tests__/Categories.test.jsx
import { render, screen } from "@testing-library/react"
import Categories from "../Categories"
import { useCategories } from "../../hooks/useProducts"
import { BrowserRouter as Router } from "react-router-dom"
import { it, expect, describe, vi } from "vitest"

// Mock the useCategories hook
vi.mock("../../hooks/useProducts", () => ({
  useCategories: vi.fn(),
}))

describe("Categories", () => {
  it("renders categories correctly", () => {
    const categories = [
      { slug: "electronics", name: "Electronics" },
      { slug: "fashion", name: "Fashion" },
    ]
    useCategories.mockReturnValue({ data: categories, isLoading: false })
    render(
      <Router>
        <Categories />
      </Router>
    )
    expect(screen.getByText(/categories/i)).toBeInTheDocument()
    categories.forEach((category) => {
      expect(screen.getByText(category.name)).toBeInTheDocument()
    })
  })

  it("renders 'View All Products' link", () => {
    useCategories.mockReturnValue({ data: [], isLoading: false })
    render(
      <Router>
        <Categories />
      </Router>
    )
    expect(screen.getByText(/view all products/i)).toBeInTheDocument()
  })

  it("renders category links correctly", () => {
    const categories = [
      { slug: "electronics", name: "Electronics" },
      { slug: "fashion", name: "Fashion" },
    ]
    useCategories.mockReturnValue({ data: categories, isLoading: false })
    render(
      <Router>
        <Categories />
      </Router>
    )
    categories.forEach((category) => {
      expect(screen.getByRole("link", { name: category.name })).toHaveAttribute(
        "href",
        `/products?category=${category.slug}`
      )
    })
  })
})
