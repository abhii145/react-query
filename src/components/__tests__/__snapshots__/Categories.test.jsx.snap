// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Categories > matches snapshot 1`] = `
<DocumentFragment>
  <section>
    <div
      class="flex justify-between items-center mb-4"
    >
      <h2
        class="text-2xl font-bold"
      >
        Categories
      </h2>
      <a
        class="text-blue-700 hover:underline text-sm"
        data-discover="true"
        href="/products"
      >
        View All Products
      </a>
    </div>
    <div
      class="grid grid-cols-2 sm:grid-cols-3 md:grid-cols-4 gap-4"
    >
      <a
        class="p-4 bg-white rounded-lg shadow hover:bg-gray-200 flex items-center gap-3"
        data-discover="true"
        href="/products?category=electronics"
      >
        <div
          class="text-blue-700 text-2xl"
        />
        <span
          class="font-medium"
        >
          Electronics
        </span>
      </a>
      <a
        class="p-4 bg-white rounded-lg shadow hover:bg-gray-200 flex items-center gap-3"
        data-discover="true"
        href="/products?category=fashion"
      >
        <div
          class="text-blue-700 text-2xl"
        />
        <span
          class="font-medium"
        >
          Fashion
        </span>
      </a>
    </div>
  </section>
</DocumentFragment>
`;
